repo/object/object_writer.go

```diff
@@ -155,12 +222,21 @@ func maybeCompressedContentBytes(comp compression.Compressor, output *bytes.Buff
 }
 
 func (w *objectWriter) Result() (ID, error) {
+	// no need to hold a lock on w.indirectIndexGrowMutex, since growing index only happens synchronously
+	// and never in parallel with calling Result()
```

Comment:
This is making an assumption about how `Result` is called w.r.t `Write(..) => `flushBuffer(...)`. It is essentially saying that index growth, and thus modifications to `w.indirectIndex`, only happen in this go routine. This works, but seems fragile. A change in the implementation outside this function, could easily result in a race here due to this assumption. Not sure what would be a good way to prevent that, and whether that type of change is already covered by unit tests and would be caught by those.

Reply:
yeah, ObjectWriter is fundamentally not goroutine-safe to the caller. tests did catch it, though, which is the reason I added the growth mutex.